docs/0002-fixed-length-n-tuple-trees.md

```diff
@@ -0,0 +1,223 @@
+* Human readability is also reduced for long lists of files, which may make recovery *in extremis* more difficult.
+
+For a **tupleSize** of 3, our example UUID of "f81d4fae7dec11d0a76500a0c91e6bf6" would be split up into a path beginning "/f81/d4f/ae7/...".
+
+A **tupleSize** of zero indicates that identifiers are not split unto tuples at all - giving a flat structure with folders 
+named for the source identifier. This can be useful for object stores that do not support the directory paradigm.
+
+### numberOfTuples
+
+In practice, you may wish to limit the depth of the OCFL Storage Hierarchy tree to avoid overly deep nesting of directories. 
+The **numberOfTuples** determines this depth. Splitting the entire identifier down into tuples may not be necessary since the
+number of objects to be stored is often much less than all the possible values for the source identifier. 
+
+For example, if we split the example UUID into size 3 tuples, each directory can contain 4096 subdirectories so, with the 
+number of tuples also set to three the resulting tree would have 4096^3 (=68719476736) directories, each of which could 
+contain one or more objects. All the objects with UUID's that begin f81d4fae7... would be stored in OCFL object roots in the
+directory /f81/d4f/ae7/. However, if the UUID's are reasonably pseudo-randomly distributed, the likelihood of many object
+identifiers sharing even the first 9 characters is quite low until a signficant number of objects have been created.
+
+The nature of the N-tuple algorithm means that the following additional conditions on parameters MUST be observed:
+* **numberOfTuples * tuplesSize <= identifierLength**
+* If **tupleSize = 0** then **numberOfTuples = 0**
+
+### shortObjectRoot
+
+Once a Storage Hierarchy has been created, the name of each OCFL Object Root directory should also be determined from the identifier in a consistent manner. The default approach is to use the full (stripped) identifier but, if identifiers are long or there is the need to keep Storage Hierarchy paths short because of object complexity, there is the option to just use the portion of the identifier that remains after Storage Hierarchy path generation. To continue the UUID example the full OCFL Object Root path could be:    
+* **shortObjectRoot = false** /f81/d4f/ae7/f81d4fae7dec11d0a76500a0c91e6bf6/
+* **shortObjectRoot = true** /f81/d4f/ae7/dec11d0a76500a0c91e6bf6/
+
+Note, if **numberOfTuples * tuplesSize = identifierLength** there is no remaining portion of the identifer and so
+**shortObjectRoot** MUST be false.
+
```

Comment:
I think we need at add a clear exposition of the algorithm using all the above parameters, rather than just relying on the example:

1. validate incoming (perhaps pre-processed) id -- characters, length
2. apply `caseMapping`
3. if `invertMapping` then reverse identifier string
4. generate `numberOfTuples` or `tupleSize` characters each as leading path segments
5. if `shortObjectRoot` then use remainder as last path segment, else use entire identifier. **If using then entire id then if might be nice to use the normal string order even if `invertMapping` specified -- either way it needs to be clear**

Reply:
Implementation Notes? Should we require this for all extensions?

Reply:
I don't think of this as something extra but as something necessary to clarify what the parameters mean and how they are used

Reply:
Agreed - but I was saying that maybe this should be required for all extensions.

Reply:
Fixed